### ISSUE TITLE
uncontrolled TextInput does not validate correctly when required and supplied with a defaultValue and used in a Form

### ISSUE THREAD

**Issue by jimwhurr** [grommet/grommet#4328]
### Expected Behavior

The input defaultValue should pass the validation for a **required** field

### Actual Behavior

The value is not accepted by the test for 'required' and results in the error message stating that the field is required.

### URL, screen shot, or Codepen exhibiting the issue
Code below is basically the example from Form on the Grommet site just with **required** attribute applied to the FormField and a **defaultValue** supplied to the TextInput

    const App = () => {

    return (
        <Grommet theme={grommet}>
        <Box pad='medium'>
            <Form onSubmit={({ value, touched }) => {console.log('value:', value, 'touched:', touched)}}>
                <FormField name='name' htmlfor='textinput-id' label='Name' required>
                    <TextInput id='textinput-id' name='name' defaultValue='Ben'/>
                </FormField>
                <Box direction='row' gap='medium'>
                    <Button type='submit' primary label='Submit' />
                    <Button type='reset' label='Reset' />
                </Box>
            </Form>
        </Box>
        </Grommet>
    )
    }

### Steps to Reproduce

1. paste code above into CRA App.js
2. npm start
3. press submit

### Your Environment

CRA app created today with grommet and styled-components:
React 16.3.1
Grommet 2.14.0
Styled Components 5.1.1

Browsers: 
Chrome 84.0.4147.89
Safari 11.1.2
Firefox 78.0.2

Tried on macos and windows 10

I also get a similar problem with a simple (username, password) login form - when Chrome password manager enters details into the fields - on submitting they give the field is required error response.

**Comment by jimwhurr** [comment]
Forgot to mention... by inserting and deleting a character in the filed it passes (e.g. put a space at the end of the default value and then delete the space - also does the same with the chrome password manager input).

**Comment by pedrosimao** [comment]
I think `defaultValue` does not existe as a prop any longer.  (https://v2.grommet.io/textinput)
The component can only be used a controlled component. In this case you would use the prop `value`...

**Comment by IanKBovard** [comment]
I agree with @pedrosimao TextInput is a controlled component. Setting a default value with state would be best here.
https://codesandbox.io/s/musing-sky-ryujg?file=/src/App.js

**Comment by jimwhurr** [comment]
Ok, I will stop using it uncontrolled.

However, if it is the case that TextInput does not support uncontrolled use, then maybe a documentation update is in order as having stories for uncontrolled use of  TextInput and other components (Form and Select at least)  is inconsistent as it suggests that to not be the case. Either way they are pretty useless if you can't set a default value in a field. 

(As an aside I thought that I read somewhere that the documentation for components does not include the definitive list of all props that are supported -- but I might be mistaken there).

Ta

Jim

**Comment by IanKBovard** [comment]
I do agree that if textinput no longer supports uncontrolled the example should be removed from storybook.I think you misunderstand, my example shows a textinput with a default value, just not using the defaultValue prop. 

I'm not too certain, but I believe the documentation refers to html attributes that the components are built upon.

**Comment by pedrosimao** [comment]
@jimwhurr 
Before using the component as a controlled one, maybe you should try passing `react-hook-form` register function to the TextInput `ref` prop. I know it is not documented, but it seems to exist.
like this:

```
<TextInput id='textinput-id' name='name' defaultValue='Ben' ref={register} />
```
where register comes from `const { register } = useForm();` of `react-hook-form`.
You can read more here:
https://react-hook-form.com/api/#register

**Comment by jimwhurr** [comment]
> I do agree that if textinput no longer supports uncontrolled the example should be removed from storybook.I think you misunderstand, my example shows a textinput with a default value, just not using the defaultValue prop.
> 
> I'm not too certain, but I believe the documentation refers to html attributes that the components are built upon.

(Sorry I didn't see your example initially as I was answering from my phone.)

That's what I tried first - think you'll find that you cannot edit the field, which was why I elected to use defaulValue in the first place. Maybe I should've reported that as the bug instead of trying to work around it?

I will change my code - the TextInput is wrapped in another component that includes the FormField as well so its fairly easy for me to make the change across the App.

Thanks

**Comment by pedrosimao** [comment]
I think if you use native `<form />`  html tag instead of Grommet's  `<Form />` and `<FormField />` it should work without problem ...

**Comment by jimwhurr** [comment]
> I think if you use native `<form />` html tag instead of Grommet's `<Form />` and `<FormField />` it should work without problem ...

Thanks for the advice :) I will try it at some point but at the moment I am on a tight timescale and I've already changed the input field to work (across all the forms) as it was only one slight code change in one file .

**Comment by IanKBovard** [comment]
@jimwhurr just checking in where we are on this issue? Were you using defaultValue because you wanted the input to be uncontrolled, or because you just wanted the input to start with an initial value? I'd like to understand your fix, so I can see if we can close this issue or not.

**Comment by pedrosimao** [comment]
@IanKBovard fir me `defautValue` is necessary for uncontrolled form components, since there is no other way to set initial values on uncontrolled components.

**Comment by IanKBovard** [comment]
@pedrosimao That makes sense to me. I think it's up the @ShimiSun if she wants to allow TextInput to be uncontrolled again, or if she wants to keep its current iteration of it being exclusively a controlled component.

**Comment by jimwhurr** [comment]
@IanKBovard 

I have changed my code to use TextInput as a comtrolled component. However if it is to exclusively be a controlled component then the examples/stories of it as uncontrolled should be removed.

If, however, you want it to be available as uncontrolled then I suggest providing defaultValue prop and ensure that Form/FormField validates it properly i.e. allows a valid, non-empty default to be unchanged to pass the required test.

Thanks

Jim :)

**Comment by IanKBovard** [comment]
it looks like you can use defaultValue
https://codesandbox.io/s/musing-sky-ryujg?file=/src/App.js:158-479s

My guess is that it does not validate is because grommet checks state, if you want to validate an uncontrolled component, i assume you have to check the dom yourself.

**Comment by jimwhurr** [comment]
I just tried using FormField with a custom Component. The docs say a component will be supported as long as long as it adheres to the API (name, value, onChange) what it doesn't say is that **it has to be controlled for it to work** (same problem as before).

Also, as an aside 
`<FormField component={Component}.../>`

works but it doesn't work if the Component is a child. To be fair though the docs don't say it will but it might be worth spelling it out.

Now I have some time I think I'll follow up on **pedrosimao's** suggestion to look at using react-hook-form as an alternative. -- Thanks :)
